### ISSUE TITLE
Primary key does not get set in upsertGraph

### ISSUE THREAD

**Issue by Bazze** [Vincit/objection.js#1556]
Hello!

I was upgrading a service of ours from an objection version of <1.5.0 to 1.6.11. However, a lot of our tests broke due to our loading of fixtures using `upsertGraph`. The issue that seems to have been introduced in version 1.5.0 is that for some reason the primary key value specified in the `upsertGraph` statement for some nested relations are not set.

I think it's easiest if you have a look at the provided reproduction code. First, run it on e.g. 1.4.0 and you should see that it's a success. Then, run it on 1.5.0 or above and you will get the test failing for sqlite or a foreign key constraint error for mysql (since it tries to insert the row in PersonAnimalToys table by referencing the `1337` id).

Is this a bug or are we misusing `upsertGraph` in some way?

Many thanks ❤️ 

```javascript
/**
 * This is a simple template for bug reproductions. It contains three models `Person`, `Animal` and `Movie`.
 * They create a simple IMDB-style database. Try to add minimal modifications to this file to reproduce
 * your bug.
 *
 * install:
 *    npm install objection knex sqlite3 chai
 *
 * run:
 *    node reproduction-template
 */

let Model;

try {
  Model = require('./').Model;
} catch (err) {
  Model = require('objection').Model;
}

const Knex = require('knex');
const chai = require('chai');

async function main() {
  await createSchema();

  ///////////////////////////////////////////////////////////////
  // Your reproduction
  ///////////////////////////////////////////////////////////////
  await Toy.query().insert({
    id: 1,
    name: 'Tennis Ball',
  });

  await Person.query().upsertGraph(
    {
      id: 10,
      firstName: 'Jennifer',
      lastName: 'Lawrence',

      pets: [
        {
          id: 1337,
          name: 'Doggo',
          species: 'dog',
          toys: [{ id: 1 }],
        },
      ],
    },
    { relate: true, unrelate: true, insertMissing: true }
  );

  const jennifer = await Person.query()
    .findOne({ firstName: 'Jennifer' })
    .eager('pets');

  chai.expect(jennifer.pets.length).to.equal(1);
  chai.expect(jennifer.pets[0].id).to.equal(1337);
  chai.expect(jennifer.pets[0].name).to.equal('Doggo');
}

///////////////////////////////////////////////////////////////
// Database
///////////////////////////////////////////////////////////////

const knex = Knex({
  client: 'sqlite3',
  useNullAsDefault: true,
  debug: false,
  connection: {
    filename: ':memory:',
  },
});

// const knex = Knex({
//   client: 'mysql',
//   connection: {
//     host: 'localhost',
//     database: 'obj_reprod',
//     user: 'root',
//     password: 'root',
//   },
// });

Model.knex(knex);

///////////////////////////////////////////////////////////////
// Models
///////////////////////////////////////////////////////////////

class Person extends Model {
  static get tableName() {
    return 'Person';
  }

  static get jsonSchema() {
    return {
      type: 'object',
      required: ['firstName', 'lastName'],

      properties: {
        id: { type: 'integer' },
        parentId: { type: ['integer', 'null'] },
        firstName: { type: 'string', minLength: 1, maxLength: 255 },
        lastName: { type: 'string', minLength: 1, maxLength: 255 },
        age: { type: 'number' },

        address: {
          type: 'object',
          properties: {
            street: { type: 'string' },
            city: { type: 'string' },
            zipCode: { type: 'string' },
          },
        },
      },
    };
  }

  static get relationMappings() {
    return {
      pets: {
        relation: Model.HasManyRelation,
        modelClass: Animal,
        join: {
          from: 'Person.id',
          to: 'Animal.ownerId',
        },
      },

      movies: {
        relation: Model.ManyToManyRelation,
        modelClass: Movie,
        join: {
          from: 'Person.id',
          through: {
            from: 'Person_Movie.personId',
            to: 'Person_Movie.movieId',
          },
          to: 'Movie.id',
        },
      },

      children: {
        relation: Model.HasManyRelation,
        modelClass: Person,
        join: {
          from: 'Person.id',
          to: 'Person.parentId',
        },
      },

      parent: {
        relation: Model.BelongsToOneRelation,
        modelClass: Person,
        join: {
          from: 'Person.parentId',
          to: 'Person.id',
        },
      },
    };
  }
}

class Animal extends Model {
  static get tableName() {
    return 'Animal';
  }

  static get jsonSchema() {
    return {
      type: 'object',
      required: ['name'],

      properties: {
        id: { type: 'integer' },
        ownerId: { type: ['integer', 'null'] },
        name: { type: 'string', minLength: 1, maxLength: 255 },
        species: { type: 'string', minLength: 1, maxLength: 255 },
      },
    };
  }

  static get relationMappings() {
    return {
      owner: {
        relation: Model.BelongsToOneRelation,
        modelClass: Person,
        join: {
          from: 'Animal.ownerId',
          to: 'Person.id',
        },
      },
      toys: {
        relation: Model.ManyToManyRelation,
        modelClass: Toy,
        join: {
          from: 'Animal.id',
          through: {
            from: 'PersonAnimalToys.animal_id',
            to: 'PersonAnimalToys.toy_id',
          },
          to: 'Toy.id',
        },
      },
    };
  }
}

class Toy extends Model {
  static get tableName() {
    return 'Toy';
  }

  static get jsonSchema() {
    return {
      type: 'object',
      required: ['name'],
      properties: {
        id: { type: 'integer' },
        name: { type: 'string', minLength: 1, maxLength: 255 },
      },
    };
  }
}

class Movie extends Model {
  static get tableName() {
    return 'Movie';
  }

  static get jsonSchema() {
    return {
      type: 'object',
      required: ['name'],

      properties: {
        id: { type: 'integer' },
        name: { type: 'string', minLength: 1, maxLength: 255 },
      },
    };
  }

  static get relationMappings() {
    return {
      actors: {
        relation: Model.ManyToManyRelation,
        modelClass: Person,
        join: {
          from: 'Movie.id',
          through: {
            from: 'Person_Movie.movieId',
            to: 'Person_Movie.personId',
          },
          to: 'Person.id',
        },
      },
    };
  }
}

///////////////////////////////////////////////////////////////
// Schema
///////////////////////////////////////////////////////////////

async function createSchema() {
  // Uncomment for mysql
  // await knex.raw('SET foreign_key_checks = 0;');

  await knex.schema
    .dropTableIfExists('Person_Movie')
    .dropTableIfExists('Animal')
    .dropTableIfExists('Movie')
    .dropTableIfExists('Person')
    .dropTableIfExists('PersonAnimalToys')
    .dropTableIfExists('Toy');

  await knex.schema
    .createTable('Person', (table) => {
      table.increments('id').primary();
      table
        .integer('parentId')
        .unsigned()
        .references('id')
        .inTable('Person');
      table.string('firstName');
      table.string('lastName');
      table.integer('age');
      table.json('address');
    })
    .createTable('Movie', (table) => {
      table.increments('id').primary();
      table.string('name');
    })
    .createTable('Animal', (table) => {
      table.increments('id').primary();
      table
        .integer('ownerId')
        .unsigned()
        .references('id')
        .inTable('Person');
      table.string('name');
      table.string('species');
    })
    .createTable('Person_Movie', (table) => {
      table.increments('id').primary();
      table
        .integer('personId')
        .unsigned()
        .references('id')
        .inTable('Person')
        .onDelete('CASCADE');
      table
        .integer('movieId')
        .unsigned()
        .references('id')
        .inTable('Movie')
        .onDelete('CASCADE');
    })
    .createTable('Toy', (table) => {
      table.increments('id').primary();
      table.string('name');
    })
    .createTable('PersonAnimalToys', (table) => {
      table
        .integer('animal_id')
        .unsigned()
        .references('id')
        .inTable('Animal');
      table
        .integer('toy_id')
        .unsigned()
        .references('id')
        .inTable('Toy');
    });

  // Uncomment for mysql
  // await knex.raw('SET foreign_key_checks = 1;');
}

main()
  .then(() => {
    console.log('success');
    return knex.destroy();
  })
  .catch((err) => {
    console.error(err);
    return knex.destroy();
  });
```

**Comment by koskimas** [comment]
This happens because the dog 1337 doesn't already have the toy 1 and therefore it gets inserted. `insertMissing` means "insert if not found in relation" not "insert if not found in database". Unfortunately this is excepted at the moment. I have no idea why it used to work. It was another bug that accidentally made it work in your case.

I would suggest avoiding upsertGraph for things that are this complex.

**Comment by koskimas** [comment]
So the problem here is that `insertMissing` means "insert if not found in relation". Objection doesn't check if the item is found in the database. It only checks the relation.

This has always been a shortcoming of `upsertGraph`. I don't know why this worked before 1.5. That was actaually a bug :smile:. This is probably the fift issue about this so I'll try to finally fix this. Fixing this will be a breaking change since people are most probably depending on the current behavior. I'll need to fix this for 2.0.

**Comment by koskimas** [comment]
This is now fixed in v2.0 branch.
